(ISSUE-9353) Fix pinned item alignment

rholinshead opened the pull request:
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes https://github.com/HabitRPG/habitica/issues/9353

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)

- Get rid of justify-content: space-between to allow flex-start default to be used
- Add margin to direct children (item-wrappers)

[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID: 9135bc92-bf74-4c15-82c2-64f1128fe950

rholinshead committed:
(ISSUE-9353) Fix pinned item alignment
- Get rid of justify-content: space-between to allow flex-start default to be used
- Add margin to direct children (item-wrappers)

paglias commented on website/client/components/tasks/column.vue:
```diff
@@ -69,7 +69,9 @@
     display: flex;
     flex-wrap: wrap;
-    justify-content: space-between;
+    > {
```
what is this selector for? maybe you meant `& > *` that selects all direct children?

rholinshead replied:
My intention (and what just using `>` seems to be doing) was to select all direct children. Since each `shopItem` is constructed with a wrapping `div` I think I'll change this to `& > div` for a more explicit selection of those `divs`. It doesn't seem like the `&` is actually needed here but if that's the convention that should be used I will stick with it.

rholinshead committed:
Issue-9353: Change selector to & > div instead of just > for more explicit selection of direct child divs

paglias commented:
LGTM but I haven't live tested the PR

rholinshead commented:
Is there some way to re-trigger the builds for this? Not sure why the tests are failing but I doubt it would be caused by this small ui change ...

paglias commented:
@rholinshead don't worry about the failing tests, we're having some problems with the build not passing right now but it's not related to your changes :)

paglias commented:
before
<img width="394" alt="schermata 2017-11-14 alle 23 35 23" src="https://user-images.githubusercontent.com/589911/32808780-8b5535ac-c994-11e7-9c3c-93e4db877226.png">

after
<img width="370" alt="schermata 2017-11-14 alle 23 35 35" src="https://user-images.githubusercontent.com/589911/32808781-8b6f2548-c994-11e7-97ec-268efbdb8650.png">

@rholinshead can you make sure the spacing remains the same?

rholinshead committed:
Merge branch 'develop' into issue-9353-pinned-item-alignment

rholinshead commented:
@paglias I will certainly try and fix the spacing so that it's closer to what it was, but there is only one way (next paragraph) to make sure it is the exact same (since it is dynamically sized due to the `justify-content`).

This is a very common issue with flexbox, and some SO answers rely on using javascript to dynamically add "blank" divs to the list of items so that the last row is always full. Personally I find that implementation very ugly due to constant calculations needed to see how many items fit in the current screen size's rewards-list (what do you think)? 

If you think that invisible divs is the way to go I can change it. Otherwise, I will spend some time cleaning up the margins to try and match the current spacing as much as possible.

paglias commented:
@rholinshead I agree with you that we should avoid blank divs if possible!

rholinshead committed:
Merge branch 'develop' into issue-9353-pinned-item-alignment

rholinshead committed:
Fix rewards item spacing to match Zeplin mockups
- Make sure horizontal/vertical spacing between items is 16px
- Add use of grid if supported, else use flex

rholinshead commented:
@paglias, sorry it's taken me so long to get back to this - have been pretty busy lately. I checked out the Zeplin design specs for the items and it shows them spaced at 16px apart. I've changed the margins here to reflect that. I also added a media query with `grid` styling so that when `grid` is supported, the item spacing looks exactly like the mockup:
![pinneditemspacing](https://user-images.githubusercontent.com/5060851/32989184-7f34b592-ccc6-11e7-995f-e48397fdb0af.png)

If `grid` isn't supported, it'll keep on using `flex` with the fixed spacing:
![pinneditemspacingflex](https://user-images.githubusercontent.com/5060851/32989225-0e7764e8-ccc7-11e7-825a-c6ba22361c65.png)


In this case, the right edge of items in the last row don't quite match up exactly with the grey "Add a Reward" box before being pushed to the next row, since doing so would require dynamically setting the margins on the items so that if they are the last one in a row they don't have the right margin. This seems like an unrealistic thing to do though and the use of `grid` will solve that

paglias commented:
@rholinshead this looks good, thanks! It'll be merged next week :)

SabreCat committed:
Merge branch 'develop' into issue-9353-pinned-item-alignment

SabreCat commented:
Doing a fresh merge of `develop` to see if it straightens out the CI failure.

SabreCat commented:
There we go!
